### PR TITLE
Compare the bitness of runtime identifier and platform target

### PR DESCRIPTION
**Pull request by livarcocc**
**Customer scenario**

User sets the project's RuntimeIdentifier to one bitness and the PlatformTarget to another. The build succeeds, but then when the user tries to run the app, it fails with a BadFormatImageException. This happens because PlatformTarget goes to the compiler, while the RuntimeIdentifier is used to determine the host. When they don't match, we end up with the host trying to load a dll that does not match its bitness. The fix is to compare the bitness of runtime identifier and platform target and error out if they don't match.

**Bugs this fixes:** 

Fixes https://github.com/dotnet/sdk/issues/698

**Workarounds, if any**

Google the exception and hopefully you will notice that you need these two properties to match.

**Risk**

Low

**Performance impact**

None.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Lack of test coverage.

**How was the bug found?**

Partner team.

@dotnet/dotnet-cli for review

@srivatsn for approval

### CODE REVIEW

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs [hunk 1]
```diff
@@ -69,5 +69,67 @@ public void It_builds_a_runnable_output()
...
+                    propertyGroup.Add(new XElement(ns + "PlatformTarget", "x86"));
+				})
+				.Restore(Log);
+
+			var buildCommand = new BuildCommand(Log, Path.Combine(testAsset.TestRoot));
+
+			buildCommand
+				.Execute()
+				.Should()
+                .Fail();
```
Comment (wli3) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs]: nit: space

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs [hunk 2]
```diff
@@ -69,5 +69,67 @@ public void It_builds_a_runnable_output()
                 .And
                 .HaveStdOutContaining("Hello World!");
         }
+
+        [Fact]
+        public void It_errors_out_when_RuntimeIdentifier_archicture_and_PlatformTarget_do_no_match()
```
Comment (wli3) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs]: nit: architecture

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs (hunk 1 above)
Comment (livarcocc) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs]: Weird, this does not show up in my editor.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs (hunk 2 above)
Comment (livarcocc) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs]: Fixed.

**Comment by wli3** [comment]
From reading the test and the change looks straight forward, but i think it is still better to have @dsplaisted have a look

**Review by wli3 — APPROVED** [review]

File: src/Tasks/Common/Resources/Strings.resx [hunk 3]
```diff
@@ -258,6 +258,9 @@
   <data name="CannotHaveSelfContainedWithoutRuntimeIdentifier" xml:space="preserve">
     <value>It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier.  Please either specify a RuntimeIdentifier or set SelfContained to false.</value>
   </data>
+  <data name="CannotHaveRuntimeIdentifierPlatformMismatchPlatformTarget" xml:space="preserve">
+    <value>The RuntimeIdentifier platform and the PlatformTarget must match.</value>
```
Comment (dsplaisted) [inline src/Tasks/Common/Resources/Strings.resx]: Saying they must match could be understood as meaning that the values need to be the same, which isn't the case.  Can we think of a better wording?

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs [hunk 4]
```diff
@@ -94,7 +94,7 @@ public void It_errors_out_when_RuntimeIdentifier_architecture_and_PlatformTarget
         }
 
 		[Fact]
-		public void It_does_not_error_out_when_RuntimeIdentifier_architecture_and_PlatformTarget_do_no_match_but_PlatformTarget_is_AnyCPU()
+		public void It_succeeds_when_RuntimeIdentifier_and_PlatformTarget_mismatch_but_PT_is_AnyCPU()
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs]: You can also pass a "methodName" parameter to the `CopyTestAsset` method if you want a descriptive method name but a shorter folder name.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets [hunk 5]
```diff
@@ -118,6 +118,15 @@ Copyright (c) .NET Foundation. All rights reserved.
     
   </Target>
 
+  <Target Name="_CheckForMismatchingPlatform"
+          BeforeTargets="_CheckForInvalidConfigurationAndPlatform"
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: When targeting .NET Framework, we default to the current architecture if it's not set, but then later back out that change.  Can you see how that interacts with this check, and add a test that covers it?

File: src/Tasks/Common/Resources/Strings.resx (hunk 3 above)
Comment (rainersigwald) [inline src/Tasks/Common/Resources/Strings.resx]: "must be compatible"?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets (hunk 5 above)
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Can you give me some more detail on this? When targeting full framework, does rid even matter?

Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: It matters, but I've looked at it more and convinced myself that extra tests for this aren't needed.

**Review by dsplaisted — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets (hunk 5 above)
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Yeah, I looked and the RuntimeIdentifierInference target and saw that we infer a win7-x64/x86 when we are full framework and a rid is not set.

**Pull request merged**